Eyes tracking

Xelkyo opened the pull request:
Stable Eyes_Tracking vizualisation, working outline.

Xelkyo committed:
Eyes Tracking vizualisation up to date

Up to the last issue comment

Xelkyo committed:
Stable Eyes_Tracking vizualisation

Eyes_Tracking vizualisation,  working outline.

bastianilso commented on Eyes_tracking/V1/app.R:
```diff
@@ -0,0 +1,66 @@
+
+
+server <- function(input, output) {
+  output$d3 <- renderD3({
+    
+    
+    r2d3(
+      data=data, 
+      script = "main.js"
+    )
+  
+    
+  })
```
Please place the preprocessing steps within `renderD3({  })`.

Xelkyo replied:
Solved, it's now in the function.

bastianilso commented on Eyes_tracking/V1/app.R:
```diff
@@ -0,0 +1,66 @@
+
+
+
+data = data %>%filter(LocalGazeDirectionX != 'NULL')
+
+
+Eyes_Data = data %>% select(LocalGazeDirectionX,LocalGazeDirectionY)
+
+
+Xtrem_points <- chull(Eyes_Data$LocalGazeDirectionX, Eyes_Data$LocalGazeDirectionY)
+plot(Eyes_Data[Xtrem_points, ])
+lines(Eyes_Data[Xtrem_points, ])
+D = Eyes_Data[Xtrem_points, ]
+
+
+
+D = D %>% mutate(is_hull = TRUE,
+  xy = paste(LocalGazeDirectionX,LocalGazeDirectionY))  %>% select(xy,is_hull)
+
+data = data %>% mutate(
+  xy = paste(LocalGazeDirectionX,LocalGazeDirectionY)
+)
+
+data = data %>% left_join(D, by = "xy")
```
These pieces of code currently exists as global variables in `app.R`. As they are preprocessing step to the D3 visualization, please keep them inside `renderD3({ })`. This is just nice custom, since it makes it clear what visualization the code belongs to.

also the following lines are probably just for debugging purposes and can be excluded or commented out with a comment `# for debugging`.
```
plot(Eyes_Data[Xtrem_points, ])
lines(Eyes_Data[Xtrem_points, ])
```

Xelkyo replied:
All useless code has been removed

bastianilso commented on Eyes_tracking/V1/main.js:
```diff
@@ -0,0 +1,191 @@
+  function storeCoords(xVal, yVal) { pointsArray.push([xVal,yVal,0])}
+  
+  
+ data.forEach((element,i) => {
+    EyesDirX.push(element.LocalGazeDirectionX)
+    EyesDirY.push(element.LocalGazeDirectionY)
+    if (element.is_hull == true) {
+    Xcoord = parseFloat(element.LocalGazeDirectionX,10)*1.02
+    Ycoord = parseFloat(element.LocalGazeDirectionY,10)*1.02
+    storeCoords(Xcoord,Ycoord)
+    }
+  }); 
```
why do we need to store LocalGazeDirectionX and Y in two separate arrays? can't we just access them from data through eg. `data["LocalGazeDirectionX"]` or something like that? Just wondering whether we can save ourselves iterating over the data manually here.

Also, is it convention in D3 code to use a for loop to perform data filtering? Please research this. Otherwise, we might benefit more from separating the data out in R and passing it via `r2d3`.

Xelkyo replied:
Solved, I know directly address like this :
```
function getCoordEyesDir(i){
   
   var Coord = [0,0];
   Coord [0] = Global_Variables.xScale(data.MainData[i].LocalGazeDirectionX*Global_Variables.zoom);
   Coord [1] = Global_Variables.yScale(data.MainData[i].LocalGazeDirectionY*Global_Variables.zoom);
   
    return([Coord[0],Coord[1]])
}
```

bastianilso reviewed (changes requested):
Hi @Xelkyo, thanks for the pull request! I have a few comments for you to start with. Some of the comments are questions that I would like you to answer in-line, within this pull request - just so I can better keep track of our conversation.

When you have addressed the comments, please make the changes inside the `Eyes_tracking branch` and push the update to github. The pull request then updates itself. Then let me know that it is ready for a second review. 👍

Xelkyo committed:
Changes according to request

I changed the R code to put all in the renderD3 function and changed the js code to reduce the for each function and directly call the data.

Please review.

bastianilso commented:
@aldsanms please see r2d3's documentation about the data structure they provide to you. From what I read, they provide the data in JSON and the data should keep their data types because they are stored via JSON conventions for encoding file types. 

more info here:
https://rstudio.github.io/r2d3/articles/data_conversion.html

Xelkyo committed:
Delete useless R treatment

As I changed the way i treat the points from the polygon I don't need anymore to had a is_hull boolean.

Xelkyo commented:
Sorry @bastianilso but didn't you switch @aldsanms pull request and mine ? Also I removed some now useless code.

bastianilso commented:
@Xelkyo sorry my bad, yea I confused the two. But the comment will apply to both your projects as it is about how the data arrives from R. 👍

bastianilso commented on Eyes_tracking/V1/app.R:
```diff
@@ -0,0 +1,51 @@
+      "hull_list" = Xtrem_points
+    )
+    
+    data_to_json <- function(data) {
+      jsonlite::toJSON(data, dataframe = "rows", auto_unbox = FALSE, rownames = TRUE)
+    } 
```
So I gather that what you are doing here is creating a custom conversion as described in https://rstudio.github.io/r2d3/articles/data_conversion.html

In that case, you may want to port over some custom arguments. For instance, in our data null values are by default encoded as "NULL" strings, whereas `jsonlite::toJSON` by default expects the value "null". You may also want to make a conversion of `Timestamp` which from the CSV is a string, but which should be mutated with `as.POSIXct()` (you can alternatively do this inside a mutate function):
https://github.com/bastianilso/WhackAMoleDataJune21/blob/main/preprocess_whackamole.R#L22

bastianilso replied:
Made a correction ^

Xelkyo replied:
Solved? I added a mutate as.POSIXct() on my data. I didn't test if it really change the format but it technically should.

bastianilso commented on Eyes_tracking/V1/main.js:
```diff
@@ -0,0 +1,186 @@
+var EyesDirX =[];
+var EyesDirY =[];
+var zoom = 1;
+var graphValueMax = 0.5;
+var pointsArray = [null]
+var pointsOnlyX = []
+var pointsOnlyY = []
```
Let's have some structure for your global variables. I think the idea of having a `graphSettings : { }` to contain things like zoom, width, height, margin, color etc is a good idea. Consult with @aldsanms on this.

Xelkyo replied:
Solved. I created an object that contains all the global variables.

bastianilso commented on Eyes_tracking/V1/main.js:
```diff
@@ -0,0 +1,186 @@
+  
+
+
+function PointLimit(x, y) {
+    this.x = x;
+    this.y = y;
+    this.angle = 0;
+}
+
+  function storeCoords(xVal, yVal) { pointsArray.push([xVal,yVal,0])}
```
I think it would also be easier to read if you organize your functions to be either in the top or at the bottom of your javascript code, so that all functions are defined first.

Then it would be nice if you organize code which currently doesnt reside in functions, within one function we can call `Main()` or `Start()`. 

You can also take inspiration from the code structure found here:
http://bl.ocks.org/nbremer/21746a9668ffdf6d8242

Xelkyo replied:
Solved. I've created a function Main() that contains all the functions calls and globally the main code, I also created a function Init() that Initialize all graphs data.

bastianilso reviewed (changes requested):
Hi @Xelkyo check out my comments. 

As for the visual, please ensure the whole HMD graphic is always within view (resizes with the window).  I've created a vector-version of the HMD graphic, which will give your graph a crisper look (non-pixelated) for this purpose:

![hmd-individual-for-vis](https://user-images.githubusercontent.com/3967945/204470595-9ed7931d-4053-459b-b701-5243894cd67c.svg)

Xelkyo committed:
Bug fix + Organization

Fixed a bug in R and organized my code in function, add an Init() function and a Main(). Init is called in the Main. Also add xScale and yScale as global variables.

Xelkyo committed:
Global variables in object

I put the global variables in an object according to the request @bastianilso made on the pull_request Eyes_Tracking

Xelkyo committed:
Fomat the time in R as POSIX

I add a mutate to make sure the time is in a right time format. It's also the last changement you ask for, can you please review my code ?

bastianilso commented:
@Xelkyo could you reply to my comments above and let me know which are solved? thanks!

bastianilso commented on Eyes_tracking/V1/main.js:
```diff
@@ -0,0 +1,174 @@
+    Global_Variables.pointsOnlyY.push(points[0][1])
+  }
+  
+function Init(){
+      svg.attr('width',900)
+    .attr('height',900)
+    .attr('style', "background: rgb(255, 255, 255);background-image: url(img/back02.png);background-repeat: no-repeat;background-position: 50% 40%    ;background-size: 140%;");
+
+    			
+    var Graph = svg.append("g")
+    .attr('width',900)
+    .attr('height',"auto")
+    
+    Global_Variables.xScale = d3.scaleLinear()
+    			.domain([Global_Variables.graphValueMax/Global_Variables.zoom, -Global_Variables.graphValueMax/Global_Variables.zoom]) 
+    			.range([0, 900]);
+    Global_Variables.yScale = d3.scaleLinear()
+    			.domain([-Global_Variables.graphValueMax/Global_Variables.zoom, Global_Variables.graphValueMax/Global_Variables.zoom]) 
+    			.range([0,900]);
+    			
+    svg.append('g')
+    	.attr('transform', 'translate(450,0)')
+    	.call(d3.axisLeft(Global_Variables.xScale).ticks(10));
+    
+    svg.append('g')
+      .attr('transform', 'translate(0,450)')
+      .call(d3.axisBottom(Global_Variables.yScale).ticks(10));   
+      
+      return Graph;
+}  
+  
+function Main(){
```
the names Init and Main might get confused. Instead of Init, maybe we need to to make it more clear `PrepareGraph` or `InitializeGraph`.
And then I would like if these functions moved to the top before any other function declaration - the declaration of Main() should come first followed by the PrepareGraph function. All functions called by Main() then appears below it.
That is, unless you have an explicit preference to do it the other way around (in that case, let me know and explain).

Xelkyo replied:
Solved, now the function is called `PrepareGraph`

bastianilso replied:
still called Main in the pull request? was the changes pushed?

Xelkyo replied:
Done and pushed

bastianilso commented on Eyes_tracking/V1/main.js:
```diff
@@ -0,0 +1,174 @@
+    }); 
+  
+      
+  data.MainData.forEach(function(e,i){
+        
+  Graph.append("circle")
+  .attr("cx", (d) => getCoordEyesDir(i)[0] )
+  .attr("cy", (d) => getCoordEyesDir(i)[1] )
+  .attr("r", 0.6)
+  .style("fill","rgb(255, 102, 102)");
+  })
```
Why do we need a `forEach` loop here? If I look at the code examples from D3's website, I don't see them using that as such (maybe im wrong). But just wondering.
See e.g. the 'Data Binding' section.

https://d3-graph-gallery.com/intro_d3js.html#data

bastianilso replied:
@Xelkyo if you could let me know about this one, thanks.

bastianilso commented on Eyes_tracking/V1/main.js:
```diff
@@ -0,0 +1,174 @@
+  .style("fill","rgb(255, 102, 102)");
+  })
+  
+  var line= d3.line()
+      .x((d,i) => getCoordEyesDir(i)[0])
+      .y((d,i) => getCoordEyesDir(i)[1])
```
using variable names like `line` and `line2` does not give much indication of what those lines actually represent. I would instead use names like `gazeLine` and `boundaryLine`. This makes it more clear for others what these lines are conceptually supposed to represent.

Xelkyo replied:
Solved, I called them `EyesPath` and `boundaryline`

bastianilso commented on Eyes_tracking/V1/main.js:
```diff
@@ -0,0 +1,174 @@
+  
+  Graph = Init();
+  
+   data.hull_list.forEach((element,i) => {              //The forEach parse a very low amount of data as it search through only hull points
+    Xcoord = parseFloat(data.MainData[data.hull_list[i]-1].LocalGazeDirectionX)*1.02      
+    Ycoord = parseFloat(data.MainData[data.hull_list[i]-1].LocalGazeDirectionY)*1.02
+    storeCoords(Xcoord,Ycoord)
+    }); 
```
Since this step isnt really drawing anything as such, maybe it could more be part of `PrepareGraph` (or if you have other suggestions)? also, this part is related to the later Sort/FindCenter etc function calls, so maybe it would make sense if the code was adjacent to each other. So that in one step you prepare the data, and then Main() will mainly concern itself with the actual `drawing`.

Xelkyo replied:
Solved ! It's now in `PrepareGraph`. I'm not sure what do you mean by adjacent to each other ?

bastianilso reviewed (changes requested):
Hi @Xelkyo, some more comments for you, have a look. ⛄

bastianilso commented on Eyes_tracking/V1/app.R:
```diff
@@ -0,0 +1,53 @@
+    )
+    
+    data_to_json <- function(data) {
+      jsonlite::toJSON(data, dataframe = "rows", auto_unbox = FALSE, rownames = TRUE)
```
why do we pass `auto_unbox = FALSE` here? just wondering

Xelkyo replied:
I genuinely doesn't understand what it does, maybe it's useless.
Here is what I found 
https://rdrr.io/cran/jsonlite/man/unbox.html

bastianilso replied:
@Xelkyo if we don't understand what it does, I think it's better to remove it and just use what is default. Explicitly stating parameters implies that we somehow don't want defaults.

Xelkyo committed:
SVG Background + variable name changes

Last stable version with SVG background and changed the names of some variables according to the Pull Request comments.

Xelkyo commented:
Pushed the SVG background and changed the variable names
![image](https://user-images.githubusercontent.com/73818268/214588411-c0ea2f92-e8bb-4361-b3f5-9ec2e93fb4f6.png)

bastianilso commented on Eyes_tracking/V1/main.js:
```diff
@@ -0,0 +1,174 @@
+  
+  Graph = PrepareGraph();
+  
+   data.hull_list.forEach((element,i) => {              //The forEach parse a very low amount of data as it search through only hull points
+    Xcoord = parseFloat(data.MainData[data.hull_list[i]-1].LocalGazeDirectionX)*1.02      
+    Ycoord = parseFloat(data.MainData[data.hull_list[i]-1].LocalGazeDirectionY)*1.02
+    storeCoords(Xcoord,Ycoord)
+    }); 
```
you say it's solved but isn't the data.hull_list forEach statement still inside DrawGraph and not inside PrepareGraph?

bastianilso commented:
@Xelkyo I added some follow-up comments, please reply to them, thanks.

Xelkyo committed:
Zoom fix

Tuesday's changes, zoom will now change everything
